File: Qytetet Java/src/controladorqytetet/ControladorQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package controladorqytetet;
import java.util.ArrayList;
import modeloqytetet.EstadoJuego;
import modeloqytetet.Qytetet;
import modeloqytetet.MetodoSalirCarcel;
/**
 *
 * @author migue
 */
public class ControladorQytetet {
    private ArrayList<String> nombreJugadores;
    private static Qytetet modelo = Qytetet.Instance();
    private static final ControladorQytetet instance = new ControladorQytetet();
    private EstadoJuego estado;
    private MetodoSalirCarcel metodo;
    
    private ControladorQytetet() {
        nombreJugadores = new ArrayList<>();
        this.estado=null;
    }
    
    public static ControladorQytetet Instance() {
        return instance;
    }
    
    public void setNombreJugadores(ArrayList<String> nombres){
        this.nombreJugadores=nombres;
        modelo.inicializarJuego(nombres);
    }
    
    public ArrayList<Integer> obtenerOperacionesJuegoValidas(){
        ArrayList<Integer> operacionesValidas=new ArrayList<>();
        this.estado= modelo.getEstadoJuego();
        
        if(nombreJugadores.isEmpty()){
            operacionesValidas.add(OpcionMenu.INICIARJUEGO.ordinal());
            return operacionesValidas;
        }
        
        switch(estado){
            case JA_CONSORPRESA:
                operacionesValidas.add(OpcionMenu.APLICARSORPRESA.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                operacionesValidas.add(OpcionMenu.TERMINARJUEGO.ordinal());
                break;
            case JA_PREPARADO:
                operacionesValidas.add(OpcionMenu.JUGAR.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                operacionesValidas.add(OpcionMenu.TERMINARJUEGO.ordinal());
                break;
            case JA_PUEDEGESTIONAR:    
                operacionesValidas.add(OpcionMenu.PASATURNO.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                operacionesValidas.add(OpcionMenu.TERMINARJUEGO.ordinal());
                if(!modelo.obtenerPropiedadesJugador().isEmpty()){
                    operacionesValidas.add(OpcionMenu.EDIFICARCASA.ordinal());
                    operacionesValidas.add(OpcionMenu.EDIFICARHOTEL.ordinal());
                    operacionesValidas.add(OpcionMenu.HIPOTECARPROPIEDAD.ordinal());
                    operacionesValidas.add(OpcionMenu.VENDERPROPIEDAD.ordinal());
                    if(!modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(true).isEmpty())
                        operacionesValidas.add(OpcionMenu.CANCELARHIPOTECA.ordinal());
                }
                break;
            case JA_PUEDECOMPRAROGESTIONAR:
                operacionesValidas.add(OpcionMenu.PASATURNO.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                operacionesValidas.add(OpcionMenu.TERMINARJUEGO.ordinal());
                operacionesValidas.add(OpcionMenu.COMPRARTITULOPROPIEDAD.ordinal());
                if(!modelo.obtenerPropiedadesJugador().isEmpty()){
                    operacionesValidas.add(OpcionMenu.EDIFICARCASA.ordinal());
                    operacionesValidas.add(OpcionMenu.EDIFICARHOTEL.ordinal()); 
                    operacionesValidas.add(OpcionMenu.VENDERPROPIEDAD.ordinal());
                    operacionesValidas.add(OpcionMenu.HIPOTECARPROPIEDAD.ordinal());
                    if(!modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(true).isEmpty())
                        operacionesValidas.add(OpcionMenu.CANCELARHIPOTECA.ordinal());
                }
                
                break;
            case ALGUNJUGADORENBANCARROTA:
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                operacionesValidas.add(OpcionMenu.TERMINARJUEGO.ordinal());
                break;
            case JA_ENCARCELADO:
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                operacionesValidas.add(OpcionMenu.TERMINARJUEGO.ordinal());
                operacionesValidas.add(OpcionMenu.PASATURNO.ordinal());
                break;
            case JA_ENCARCELADOCONOPCIONDELIBERTAD:
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORACTUAL.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARJUGADORES.ordinal());
                operacionesValidas.add(OpcionMenu.MOSTRARTABLERO.ordinal());
                operacionesValidas.add(OpcionMenu.TERMINARJUEGO.ordinal());
                operacionesValidas.add(OpcionMenu.INTENTARSALIRCARCELPAGANDOLIBERTAD.ordinal());
                operacionesValidas.add(OpcionMenu.INTENTARSALIRCARCELTIRANDODADO.ordinal());
        }
        return operacionesValidas;
        
    }

    public boolean necesitaElegirCasilla(int opcionMenu){
        boolean resultado=false;
        if(opcionMenu==OpcionMenu.HIPOTECARPROPIEDAD.ordinal()||opcionMenu==OpcionMenu.CANCELARHIPOTECA.ordinal()
                || opcionMenu==OpcionMenu.EDIFICARCASA.ordinal() || opcionMenu==OpcionMenu.EDIFICARHOTEL.ordinal()
                || opcionMenu==OpcionMenu.VENDERPROPIEDAD.ordinal())
            resultado=true;
        return resultado;
    }
    
    public ArrayList<Integer> obtenerCasillasValidas(int opcionMenu){
        ArrayList<Integer> validas = new ArrayList<>();
        if(opcionMenu == OpcionMenu.EDIFICARCASA.ordinal() || opcionMenu == OpcionMenu.EDIFICARHOTEL.ordinal() 
                || opcionMenu == OpcionMenu.VENDERPROPIEDAD.ordinal()){
            validas=modelo.obtenerPropiedadesJugador();
        }
        else{
            if(opcionMenu==OpcionMenu.CANCELARHIPOTECA.ordinal()){
                validas=modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(true);
            }
            else{
                if(opcionMenu == OpcionMenu.HIPOTECARPROPIEDAD.ordinal())
                    validas=modelo.obtenerPropiedadesJugadorSegunEstadoHipoteca(false);
            }
        }
        return validas;
    }
    
    public String realizarOperacion(int opcionElegida, int casillaElegida){
        String devolver=null;
        OpcionMenu opcion=OpcionMenu.values()[opcionElegida];
        switch(opcion){
            case INICIARJUEGO:
                modelo.inicializarJuego(nombreJugadores);
                devolver="Juego iniciado";
                break;
            case JUGAR:
                modelo.jugar();
                devolver="El valor que ha salido en el dado es "+modelo.getValorDado()+". Has caido en la casilla"+modelo.obtenerCasillaJugadorActual();
                break;
            case APLICARSORPRESA:
                modelo.aplicarSorpresa();
                devolver="SORPRESA: "+modelo.getCartaActual();
                break;
            case INTENTARSALIRCARCELPAGANDOLIBERTAD:
                boolean salir=false;
                salir=modelo.intentarSalirCarcel(MetodoSalirCarcel.PAGANDOLIBERTAD);
                if(!salir)
                    devolver="Pagas la fiaza. ¡Sales de la cárcel!";
                else
                    devolver="No te puedes permitir pagar la fianza, sigues en la cárcel";
                break;
            case INTENTARSALIRCARCELTIRANDODADO:
                boolean salir_dado=false;
                salir_dado=modelo.intentarSalirCarcel(MetodoSalirCarcel.TIRANDODADO);
               
                if(!salir_dado)
                    devolver="Has sacado un "+modelo.getValorDado()+". Es tu día de suerte, ¡Sales de la cárcel!";
                else
                    devolver="Has sacado un "+modelo.getValorDado()+".Otra vez será, pero de momento te quedas en la cárcel";
                break;
            case COMPRARTITULOPROPIEDAD:
                boolean comprado=modelo.comprarTituloPropiedad();
                if(comprado)
                    devolver="¡Enhorabuena! acabas de adquirir la propiedad de la calle "+modelo.obtenerCasillaJugadorActual();
                else
                    devolver="No es posible comprar la propiedad de la casilla actual...";
                break;
            case HIPOTECARPROPIEDAD:
                modelo.hipotecarPropiedad(casillaElegida);
                devolver="Se ha hipotecado la propiedad de la casilla "+casillaElegida;
                break;
            case CANCELARHIPOTECA:
                boolean cancelada=modelo.cancelarHipoteca(casillaElegida);
                if(cancelada)
                    devolver="Se ha cancelado con éxito la casilla "+casillaElegida;
                else
                    devolver="No ha sido posible cancelar la hipoteca de la casilla "+casillaElegida+"...";
                break;
            case EDIFICARCASA:
                boolean casa_edificada=modelo.edificarCasa(casillaElegida);
                if(casa_edificada)
                    devolver="Has construido una casa en la casilla "+casillaElegida;
                else
                    devolver="No ha sido posible edificar la casa en la casilla "+casillaElegida+"...";
                break;
            case EDIFICARHOTEL:   
                boolean hotel_edificado=modelo.edificarHotel(casillaElegida);
                if(hotel_edificado)
                    devolver="Has construido un hotel en la casilla "+casillaElegida;
                else
                    devolver="No ha sido posible edificar el hotel en la casilla "+casillaElegida+"...";
                break;
            case VENDERPROPIEDAD:
                modelo.venderPropiedad(casillaElegida);
                devolver="Has vendido la propiedad de la casilla "+casillaElegida;
                break;
            case PASATURNO:
                modelo.siguienteJugador();
                devolver="El jugador actual termina su turno";
                break;
                
            case TERMINARJUEGO:
                System.out.println("El ranking es: ");
                modelo.obtenerRanking();
                System.out.println(modelo.getJugadores());
                System.out.println("Fin del juego");
                System.exit(0);
                break;
            case MOSTRARJUGADORACTUAL:
                System.out.println(modelo.getJugadorActual());
                break;
            case MOSTRARJUGADORES:
                System.out.println(modelo.getJugadores());
                break;
            case MOSTRARTABLERO:
                System.out.println(modelo.getTablero());
                break;
            default:
                break;
        }
        return devolver;
    }

}

File: Qytetet Ruby/lib/controlador_qytetet.rb
#encoding: utf-8
require "singleton"
require_relative "estado_juego"
require_relative "qytetet"
require_relative "opcion_menu"
require_relative "metodo_salir_carcel"
module ControladorQytetet
  class ControladorQytetet
    include Singleton
    @@modelo=ModeloQytetet::Qytetet.instance 
    def initialize
      @nombre_jugadores=Array.new
      @estado=nil
      @metodo=nil
    end
    
    def set_nombre_jugadores(nombres)
      @nombre_jugadores=nombres
      @@modelo.inicializar_juego(@nombre_jugadores)
    end
    
    def obtener_operaciones_juego_validas
      operaciones_validas=Array.new
      @estado=@@modelo.estado
      
      if @nombre_jugadores.empty?
        operaciones_validas<<OpcionMenu.index(:INICIARJUEGO)
        return operaciones_validas
      end
      
      case @@modelo.estado
      when ModeloQytetet::EstadoJuego::JA_CONSORPRESA
        operaciones_validas<<OpcionMenu.index(:APLICARSORPRESA)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORACTUAL)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORES)
        operaciones_validas<<OpcionMenu.index(:MOSTRARTABLERO)
        operaciones_validas<<OpcionMenu.index(:TERMINARJUEGO)
      when ModeloQytetet::EstadoJuego::JA_PREPARADO
        operaciones_validas<<OpcionMenu.index(:JUGAR)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORACTUAL)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORES)
        operaciones_validas<<OpcionMenu.index(:MOSTRARTABLERO)
        operaciones_validas<<OpcionMenu.index(:TERMINARJUEGO)
      when ModeloQytetet::EstadoJuego::JA_PUEDEGESTIONAR
        operaciones_validas<<OpcionMenu.index(:PASARTURNO)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORACTUAL)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORES)
        operaciones_validas<<OpcionMenu.index(:MOSTRARTABLERO)
        operaciones_validas<<OpcionMenu.index(:TERMINARJUEGO)
        if !@@modelo.obtener_propiedades_jugador.empty?
          operaciones_validas<<OpcionMenu.index(:EDIFICARCASA)
          operaciones_validas<<OpcionMenu.index(:EDIFICARHOTEL)
          operaciones_validas<<OpcionMenu.index(:HIPOTECARPROPIEDAD)
          operaciones_validas<<OpcionMenu.index(:VENDERPROPIEDAD)
          if !@@modelo.obtener_propiedades_jugador_segun_estado_hipoteca(true).empty?
            operaciones_validas<<OpcionMenu.index(:CANCELARHIPOTECA)
          end
        end
      when ModeloQytetet::EstadoJuego::JA_PUEDECOMPRAROGESTIONAR
        operaciones_validas<<OpcionMenu.index(:PASARTURNO)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORACTUAL)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORES)
        operaciones_validas<<OpcionMenu.index(:MOSTRARTABLERO)
        operaciones_validas<<OpcionMenu.index(:TERMINARJUEGO)
        operaciones_validas<<OpcionMenu.index(:COMPRARTITULOPROPIEDAD)
        if !@@modelo.obtener_propiedades_jugador.empty?
          operaciones_validas<<OpcionMenu.index(:EDIFICARCASA)
          operaciones_validas<<OpcionMenu.index(:EDIFICARHOTEL)
          operaciones_validas<<OpcionMenu.index(:HIPOTECARPROPIEDAD)
          operaciones_validas<<OpcionMenu.index(:VENDERPROPIEDAD)
          if !@@modelo.obtener_propiedades_jugador_segun_estado_hipoteca(true).empty?
            operaciones_validas<<OpcionMenu.index(:CANCELARHIPOTECA)
          end
        end
      when ModeloQytetet::EstadoJuego::ALGUNJUGADORENBANCARROTA
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORACTUAL)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORES)
        operaciones_validas<<OpcionMenu.index(:MOSTRARTABLERO)
        operaciones_validas<<OpcionMenu.index(:TERMINARJUEGO)
      when ModeloQytetet::EstadoJuego::JA_ENCARCELADO
        operaciones_validas<<OpcionMenu.index(:PASARTURNO)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORACTUAL)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORES)
        operaciones_validas<<OpcionMenu.index(:MOSTRARTABLERO)
        operaciones_validas<<OpcionMenu.index(:TERMINARJUEGO)
      when ModeloQytetet::EstadoJuego::JA_ENCARCELADOCONOPCIONDELIBERTAD
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORACTUAL)
        operaciones_validas<<OpcionMenu.index(:MOSTRARJUGADORES)
        operaciones_validas<<OpcionMenu.index(:MOSTRARTABLERO)
        operaciones_validas<<OpcionMenu.index(:TERMINARJUEGO)
        operaciones_validas<<OpcionMenu.index(:INTENTARSALIRCARCELPAGANDOLIBERTAD)
        operaciones_validas<<OpcionMenu.index(:INTENTARSALIRCARCELTIRANDODADO)
      end
      return operaciones_validas
    end
    
    def necesita_elegir_casilla(opcion_menu)
      resultado=false
      if (opcion_menu == OpcionMenu.index(:HIPOTECARPROPIEDAD) || opcion_menu == OpcionMenu.index(:CANCELARHIPOTECA) || 
            opcion_menu == OpcionMenu.index(:EDIFICARCASA) || opcion_menu == OpcionMenu.index(:EDIFICARHOTEL) || 
            opcion_menu == OpcionMenu.index(:VENDERPROPIEDAD))
        resultado=true
      end
      return resultado
    end
    
    def obtener_casillas_validas(opcion_menu)
      validas=Array.new
      if (opcion_menu == OpcionMenu.index(:EDIFICARCASA) || opcion_menu == OpcionMenu.index(:Edificar_hotel) || 
            opcion_menu == OpcionMenu.index(:VENDERPROPIEDAD))
        validas = @@modelo.obtener_propiedades_jugador
      else
        if opcion_menu == OpcionMenu.index(:CANCELARHIPOTECA)
          validas = @@modelo.obtener_propiedades_jugador_segun_estado_hipoteca(true)
        else
          if opcion_menu == OpcionMenu.index(:HIPOTECARPROPIEDAD)
            validas = @@modelo.obtener_propiedades_jugador_segun_estado_hipoteca(false)
          end
        end
      end
 
    return validas
    end
    
    def realizar_operacion(opcion_elegida,casilla_elegida)
      case opcion_elegida
      when OpcionMenu.index(:INICIARJUEGO) 
        @@modelo.inicializar_juego(@nombre_jugadores)
        devolver="Juego iniciado"
      when OpcionMenu.index(:JUGAR) 
        @@modelo.jugar 
        devolver="El valor que ha salido en el dado es " + @@modelo.get_valor_dado.to_s + ". Has caido en la casilla "+@@modelo.obtener_casilla_jugador_actual.to_s
      when OpcionMenu.index(:APLICARSORPRESA)
        @@modelo.aplicar_sorpresa
        devolver="SORPRESA: " + @@modelo.carta_actual.to_s
      when OpcionMenu.index(:INTENTARSALIRCARCELPAGANDOLIBERTAD)
        salir = @@modelo.intentar_salir_carcel(ModeloQytetet::MetodoSalirCarcel::PAGANDOLIBERTAD)
        if salir
          devolver="Pagas la fianza. ¡Sales de la cárcel!"
        else
          devolver="No te puedes permitir pagar la fianza, sigues en la cárcel"
        end
      when OpcionMenu.index(:INTENTARSALIRCARCELTIRANDODADO)
        salir_dado = @@modelo.intentar_salir_carcel(ModeloQytetet::MetodoSalirCarcel::TIRANDODADO)
        if salir_dado
          devolver="Has sacado un " + @@modelo.get_valor_dado.to_s + ". Es tu día de suerte, ¡Sales de la cárcel!"
        else
          devolver="Has sacado un " + @@modelo.get_valor_dado.to_s + ". Otra vez será, pero de momento te quedas en la cárcel"
        end
      when OpcionMenu.index(:COMPRARTITULOPROPIEDAD)
        comprado = @@modelo.comprar_titulo_propiedad
        if comprado
          devolver="¡Enhorabuena! acabas de adquirir la propiedad de la calle" + @@modelo.obtener_casilla_jugador_actual.to_s
        else
          devolver="No es posible comprar la propiedad de la casilla actual..."
        end
      when OpcionMenu.index(:HIPOTECARPROPIEDAD)
        @@modelo.hipotecar_propiedad(casilla_elegida)
        devolver="Se ha hipotecado la propiedad de la casilla " + casilla_elegida.to_s
      when OpcionMenu.index(:CANCELARHIPOTECA)
        cancelada = @@modelo.cancelar_hipoteca(casilla_elegida)
        if cancelada
          devolver="Se ha cancelado con éxito la casilla " + casilla_elegida.to_s
        else
          devolver="No ha sido posible cancelar la hipoteca de la casilla " + casilla_elegida.to_s
        end
      when OpcionMenu.index(:EDIFICARCASA)
        casa_edificada = @@modelo.edificar_casa(casilla_elegida)
        if casa_edificada
          devolver="Has construido una casa en la casilla " + casilla_elegida.to_s 
        else
          devolver="No ha sido posible edificar la casa en la casilla " + casilla_elegida.to_s
        end
      when OpcionMenu.index(:EDIFICARHOTEL)
        casa_edificada = @@modelo.edificar_hotel(casilla_elegida)
        if casa_edificada
          devolver="Has construido un hotel en la casilla " + casilla_elegida 
        else
          devolver="No ha sido posible edificar el hotel en la casilla" + casilla_elegida
        end
      when OpcionMenu.index(:VENDERPROPIEDAD) 
        @@modelo.vender_propiedad(casilla_elegida)
        devolver="Has vendido la propiedad de la casilla " + casilla_elegida
      when OpcionMenu.index(:PASARTURNO)
        @@modelo.siguiente_jugador
        devolver = "El jugador actual termina su turno"
      when OpcionMenu.index(:TERMINARJUEGO)
        @@modelo.obtener_ranking
        puts "El ranking es: "
        puts @@modelo.jugadores.to_s
        puts "Fin del juego"
        exit(0)
      when OpcionMenu.index(:MOSTRARJUGADORACTUAL)
        puts @@modelo.jugador_actual.to_s
      when OpcionMenu.index(:MOSTRARJUGADORES)
        puts @@modelo.jugadores.to_s
      when OpcionMenu.index(:MOSTRARTABLERO) 
        puts @@modelo.tablero.to_s
        
      end

      return devolver
    end
  end    
end
